### ISSUE TITLE
Stop NeptuneLogger experiment on deallocation

### ISSUE THREAD

**Issue by fmorlock-tt** [Lightning-AI/lightning#15866]
### Bug description

The [NeptuneLogger](https://pytorch-lightning.readthedocs.io/en/stable/_modules/pytorch_lightning/loggers/neptune.html) class does not appear to stop the experiment tracking (leaving a zombie connection to the logger server) when deallocated.

When a training session in a script running multiple training sessions ends, the corresponding logging session to neptune.ai remain open. I've encountered this while playing around with Optuna and Raytune doing hyperparameters sweeps in a multiprocessing environment. 

I'm not sure the proper way to kill the connection from within the `NeptuneLogger` class, however, adding this destructor seems to work to _stop_ the experiment. I think it might also fix some other situations where the neptune logger backend does not disconnect despite the PT logger going out of scope.

```python
def __del__(self):
    self._run_instance.stop()
```

I hope everyone doesn't mind, but I'll loop in the two people I know over at neptune.ai (Hi @Raalsky and @Blaizzy).

### How to reproduce the bug

My situation is a bit different, but if you have a script that runs multiple training sessions in parallel, then the instances that finish first will remain open and "logging" despite the training having exited and that training instance code being inaccessible.

EDIT: See reproduction code here https://github.com/Lightning-AI/lightning/issues/15866#issuecomment-1332559008

### Error messages and logs

No errors except for my logging hours being eaten up 😢 

### Environment

[Full env here https://pastebin.com/msgWm587](https://pastebin.com/msgWm587)

cc @Blaizzy

**Comment by Blaizzy** [comment]
Hey @fmorlock-tt 

Thanks for keeping us in the loop

I just tried to reproduce the bug by creating multiple runs using a script but Neptune is behaving normally. All the runs get the kill signal after and stop on time. You can verify this by checking the monitoring time and Run time.

On the other hand, I noticed that the spinning icon takes a few seconds to refresh and reflect the actual state of the backend. In such a case, you can just manually refresh the website.

Finally, it worked on my side but you might be experiencing an actual bug, so let's run a simple test.

Please run this script and send me the `sys/monitoring_time` values for each.

```python
import neptune.new as neptune

for _ in range(3):
    run = neptune.init_run()

```

**Comment by fmorlock-tt** [comment]
Hey @Blaizzy 

I ran a modification to your code (with a sleep timer) to show the issue better. You can see that the other instances are kept alive for the entire lifespan of the script.
![image](https://user-images.githubusercontent.com/111152750/204874919-1675b58d-53cb-4a8d-9b51-4ebe19ff5260.png)
```python
import neptune.new as neptune
import time

with open('.neptune_config', 'r') as f:
    api_key = f.read().strip()

for _ in range(3):
    run = neptune.init_run(api_token=api_key,
            project='workspace/project',
            name=f"BAD MODEL",)
    time.sleep(10)
```


Here is some code I wrote up this morning to demonstrate the behavior in PyTorch Lightning

```python
import torch
from pytorch_lightning import LightningModule, Trainer
from torch import nn
from torch.nn import functional as F
from torch.utils.data import DataLoader
from torchvision import transforms
from torchvision.datasets import MNIST
from pytorch_lightning.loggers import NeptuneLogger

import multiprocess as mp

with open('.neptune_config', 'r') as f:
    api_key = f.read().strip()

# this dataset doesn't play well with multiprocessing -- thus it's initialized globally
train_ds = MNIST(f'./mnist', train=True, download=False, transform=transforms.ToTensor())

class MNISTModel(LightningModule):
    def __init__(self):
        super().__init__()
        self.l1 = torch.nn.Linear(28 * 28, 10)

    def forward(self, x):
        return torch.relu(self.l1(x.view(x.size(0), -1)))

    def training_step(self, batch, batch_nb):
        x, y = batch
        loss = F.cross_entropy(self(x), y)
        self.log('metrics/train/loss', loss)
        return loss

    def configure_optimizers(self):
        return torch.optim.Adam(self.parameters(), lr=0.02)

def train(epochs: int):
    print(f'training for {epochs} epochs')
    # Init our model
    mnist_model = MNISTModel()
    # Init DataLoader from MNIST Dataset (for some reason it doesn't like from read from the same dir -- hands)
    train_loader = DataLoader(train_ds, batch_size=64)
    
    logger = NeptuneLogger(
            api_key=api_key,
            project='workspace/project',
            name=f"BAD MODEL",
            tags=["debugging"],
            source_files=["trainer.py"],
            log_model_checkpoints=False,
            # mode='debug'
        )

    # Initialize a trainer
    trainer = Trainer(
        accelerator="gpu",
        max_epochs=epochs,
        logger=logger
    )

    # Train the model ⚡
    trainer.fit(mnist_model, train_loader)
    print(f'finished training for {epochs}')

if __name__ == '__main__':
    epoch_grid = [5, 10, 20]

    print('Starting')
    # both multiprocessing and single-threaded code keeps the connections open
    # with mp.Pool(4) as p:
    #     p.map(train, epoch_grid)
    for epoch in epoch_grid:
        train(epoch)
    print('Done')

```

With multiprocessing (last column is number of epochs completed)
![image](https://user-images.githubusercontent.com/111152750/204876343-4641b5a2-abe1-49ec-b377-04aed23569d4.png)

Without multiprocessing
![image](https://user-images.githubusercontent.com/111152750/204876414-659ea371-f3db-47a1-bc1a-f13c53ae80ee.png)

**Comment by Blaizzy** [comment]
Thanks for the detailed explanation.

You can manually stop a `run` object by using the `stop()` method. 

First, use the PTL NeptuneLogger instance attribute `experiment` to retrieve the `run` object then call the stop() method to stop it.

Example:
```python
import torch
from pytorch_lightning import LightningModule, Trainer
from torch import nn
from torch.nn import functional as F
from torch.utils.data import DataLoader
from torchvision import transforms
from torchvision.datasets import MNIST
from pytorch_lightning.loggers import NeptuneLogger

import multiprocess as mp

with open('.neptune_config', 'r') as f:
    api_key = f.read().strip()

# this dataset doesn't play well with multiprocessing -- thus it's initialized globally
train_ds = MNIST(f'./mnist', train=True, download=False, transform=transforms.ToTensor())

class MNISTModel(LightningModule):
    def __init__(self):
        super().__init__()
        self.l1 = torch.nn.Linear(28 * 28, 10)

    def forward(self, x):
        return torch.relu(self.l1(x.view(x.size(0), -1)))

    def training_step(self, batch, batch_nb):
        x, y = batch
        loss = F.cross_entropy(self(x), y)
        self.log('metrics/train/loss', loss)
        return loss

    def configure_optimizers(self):
        return torch.optim.Adam(self.parameters(), lr=0.02)

def train(epochs: int):
    print(f'training for {epochs} epochs')
    # Init our model
    mnist_model = MNISTModel()
    # Init DataLoader from MNIST Dataset (for some reason it doesn't like from read from the same dir -- hands)
    train_loader = DataLoader(train_ds, batch_size=64)
    
    logger = NeptuneLogger(
            api_key=api_key,
            project='workspace/project',
            name=f"BAD MODEL",
            tags=["debugging"],
            source_files=["trainer.py"],
            log_model_checkpoints=False,
            # mode='debug'
        )

    # Initialize a trainer
    trainer = Trainer(
        accelerator="gpu",
        max_epochs=epochs,
        logger=logger
    )

    # Train the model ⚡
    trainer.fit(mnist_model, train_loader)
    print(f'finished training for {epochs}')
    
   # Stop neptune run
   logger.experiment.stop() 


if __name__ == '__main__':
    epoch_grid = [5, 10, 20]

    print('Starting')
    # both multiprocessing and single-threaded code keeps the connections open
    # with mp.Pool(4) as p:
    #     p.map(train, epoch_grid)
    for epoch in epoch_grid:
        train(epoch)
    print('Done')

```

Docs: 
- https://docs.neptune.ai/usage/best_practices/#stopping-runs-and-other-objects
- https://github.com/neptune-ai/examples/blob/main/integrations-and-supported-tools/pytorch-lightning/notebooks/Neptune_PyTorch_Lightning.ipynb

**Comment by Blaizzy** [comment]
Let me know if this works for you

**Comment by fmorlock-tt** [comment]
I am proposing utilizing the garbage collector (via a [destructor `__del__`](https://docs.python.org/3/reference/datamodel.html#object.__del__)) to stop/clean up the would-be zombie logging processes when the NeptuneLogger is no longer referenced by anything and is garbage collected. The code I posted above is not my project code, but rather an example of how the logger can persist beyond it's reasonable lifetime. I know that you can manually stop the logging server, in much the same way that you can manually `free` allocated memory, however I believe that it's important to make use of the garbage collector to ensure those resources are freed when the container object is.

Again, **my particular use case is slightly different than the code I posted**. I am using a [hyperparameter search scheduler](https://docs.ray.io/en/latest/tune/api_docs/schedulers.html#asha-tune-schedulers-ashascheduler) that will train models in parallel and will kill instances that aren't performing well. Because the logging backend is not cleaned up when the `NeptuneLogger` is deallocated (when a run is killed), I have connection to Neptune (for each run) that can only be closed by killing the main process, ending all my runs. While my knowledge of [Ray Tune](https://docs.ray.io/en/latest/tune/index.html) process management systems is not complete, the job manager attempts to clean up the resources of a process when shutting down. The neptune.ai logging backend does not get closed and but *can* be automatically closed when the `NeptuneLogger` instance is destroyed if a destructor/finalizer is used. 

Is there a reason why you would be against using the garbage collector? The lifetime of the containing object has ended, however, the logging server owned by that object does not get killed. I would consider this something akin to a memory leak. If there is no reason to persist a logging server is the background, then I would propose that **the lifecycle of an acquired resource should be bound to the lifetime of the acquiring object** (ala [RAII principle](https://en.cppreference.com/w/cpp/language/raii)).

**Comment by Blaizzy** [comment]
I understand @fmorlock-tt, and I see your point.

I will submit it as a feature request,

Additionally, I'm not against it. Instead, I want to understand your use case better. 

Here is why:

1.  The impact of adding this feature to other use cases beyond the one you presented needs to be clarified, given that the PTL integration functions correctly and there is a way to stop runs. 

2. I believe this feature request is for HPO framework integration, such as Optuna, which we currently have available, and RayTune integration which we already have in our backlog. Instead of the PTL integration.


I'll need to investigate it further and talk to the team to hear what they say. I'll keep you updated.

Meanwhile, you can use the following solution so you can continue your work: 
-  Optuna guide that demonstrates how to log trials as separate Neptune runs. Docs: https://docs.neptune.ai/integrations/optuna/#logging-each-trial-as-separate-neptune-run

- If you prefer RayTune, you can use the stop method. Docs: https://docs.neptune.ai/usage/best_practices/#stopping-runs-and-other-objects

**Comment by Blaizzy** [comment]
Hi @fmorlock-tt 

I just heard from the team and they too suggest using the `stop()` method or optionally you can customize our logger to add the destructor.

I also passed your feedback on to the product team for future reference and consideration.

Let me know if this works for you.

**Comment by Blaizzy** [comment]
Hey @fmorlock-tt 

Just checking in to see if you still need help with this or if you need help with anything else. I'm happy to help:)
